### DISCUSSION TITLE
Windows Terminal can only be installed in user scope

### DISCUSSION THREAD

**Discussion by cheesecracker** [microsoft/winget-cli#1983] [Q&A]
### Brief description of your issue

Hello,

Currently, Windows Terminal can ONLY be installed in user scope.  That means:

* If I run the command "winget install -e --id Microsoft.WindowsTerminal", only current user can use the application
* If I run the command "winget install --scope machine -e --id Microsoft.WindowsTerminal", a red error message is returned and the application is NOT installed

Windows Terminal is the only application I find to have this behaviour.  Please tell me how to install it in machine scope, if possible.

Thanks

### Steps to reproduce

Run the command "winget install --scope machine -e --id Microsoft.WindowsTerminal"

### Expected behavior

Application is installed in machine scope

### Actual behavior

RED error message and the application is not installed

### Environment

```shell
Windows Package Manager v1.1.13405
Copyright (c) Microsoft Corporation. Tous droits réservés.

Windows: Windows.Desktop v10.0.19044.1566
Package: Microsoft.DesktopAppInstaller v1.16.13405.0
```

**Comment by ItzLevvie** [comment]


**Reply by ghost** [reply]
> If you would like to deploy an MSIX app system-wide instead of the current user

Nope. To this date, MSIX does _**NOT**_ support "SYSTEM WIDE / PER MACHINE INSTALLATION" _at all_.

 👉 Machine Wide Provisioning ( aka `Add-AppxProvisionedPackage` ) still **_installs/registers_** the app on per user basis. 

See Feature Requests :  
1. https://github.com/microsoft/WindowsAppSDK/issues/13
2. https://techcommunity.microsoft.com/t5/msix-feedback/machine-wide-provisioning-install-for-all-users/idi-p/1781716

**Comment by eabase** [comment]
@ItzLevvie 
I'm not sure I understand this right. Can you reformulate why it is not possible to install via a specified scope , when indeed you have to command option `--scope machine` available?

I.e. What is the purpose of the `--scope` option then?

**Reply by florelis** [reply]
The `--scope` options does specify whether to install machine-wide or for the current user only. However not all applications support being installed in both scopes. In some cases, like Windows Terminal, it is because of the application's install technology not supporting it, like MSIX. It could also be that the publisher decides not to support both scopes, or simply that we don't have the right information in the package's manifest.

**Comment by cheesecracker** [comment]
> This is the intended behavior of how MSIX apps are installed by default.
> 
> This applies to all MSIX apps—not just for Windows Terminal—as all MSIX apps are only installed for the current user even if you didn't install it through the Windows Package Manager.

I see.  That's indeed sad (and lame).  Installing applications in user scope but impossible in machine scope is retrograding to the 90's!  And it's not suitable for enterprise environment.  This software is thus ONLY good for individuals.

By curiosity: if the author did not choose MSIX, it could have been capable of being machine-wide?

**Reply by denelon** [reply]
Both MSI and .exe-based installers can be built to support both machine and user scope, but it depends on the package author.

**Reply by cheesecracker** [reply]
@denelon
So, you mean "both MSI and .exe-based installers can be both machine and user scope, while MSIX is specifically user scope only"?  Did I get you right?

**Reply by denelon** [reply]
MSIX was designed to address several different concerns and behaves differently from a traditional .exe or .msi based installer. MSIX packages have a different lifecycle and when the intent is to be "installed" for each user on the system (the binaries aren't duplicated), but the virtual registry entries aren't created until the user interactively logs in. MSIX was designed to be installed and executed using a form of isolation or containerization. 

https://learn.microsoft.com/windows/msix/overview

They can be installed "logically" machine-wide, but there is a subtle difference in mechanically how that is achieved.

**Reply by cheesecracker** [reply]
@denelon 
OK, thanks for the detailed explanation.  That's more than I expected.  I really appreciate it.

Now, enough of the technical side.  Back to the practical side / usefulness side of this matter.  So I suppose there is NO WAY to install this application in machine-wide in company computers.  That means the **only solution** for me is to abandon this application and try to find something similar, right?

Do people here have any suggestion of any alternative to this application, please?

**Reply by denelon** [reply]
We've been doing work to support WinGet in the system context. This is likely to be limited to the PowerShell module, and we will need to change the behavior used for installing MSIX packages in this scope. It is technically possible to have them available for users on their first login, but we will have to modify the Install-WinGetPackage in the system context to make this possible.

**Reply by cheesecracker** [reply]
A little sidenote:

I don't know if the following fact is related to this issue of the matter or not.

Windows Terminal installed from **Microsoft Store** is also user-wide only.

**Reply by discotimetraveler** [reply]
As someone who ran a Citrix / Terminal Services environment for 20 years, this is simply an absurd decision from Microsoft. It's like the people who came up with this started using a computer last week.

**Reply by denelon** [reply]
@discotimetraveler I'm not sure what is meant by "absurd decision". If you can clarify, maybe I could shed some light on this concern.

**Reply by cheesecracker** [reply]
> @discotimetraveler I'm not sure what is meant by "absurd decision". If you can clarify, maybe I could shed some light on this concern.

Windows Terminal, which is very technical, is most likely used a lot in company / professional environment.  And in company / professional environment, we have to consider that *every* computer is meant to be used by any company employee.

Let me be clear as an IT admin working over 20 years in pro.  When we prepare a new computer, we have to consider that it's meant to be *generic*.  We have no idea who is going to use such and such computer.  So when we install software/application in a new computer, we have to make sure that *any* user logging into it will see the same things as anybody else.  And we have to make sure that new user would spend the *less time possible* on the first log-in.  User/employee time is precious.  Wasting employee time means wasting company time.

Hope this is clear.

**Reply by denelon** [reply]
@cheesecracker,
Thanks for the clarification. I'll try to be a little bit more explicit here myself.

MSIX packages can be installed in the system context. There is just a unique lifecycle associated with an MSIX package that isn't the same as an MSI or a legacy .exe-based install. The notion of "machine wide" for an MSIX based package is handled differently in the sense that the virtual registry entries and the container the packaged application runs in doesn't get provisioned for the user until they log in.

MSIX packages are logically "added to the path" via the App Execution Alias. Given WinGet is shipped inside the App Installer (a packaged application), attempting to run `winget` in the system context will fail because the App Execution Alias concept was intended for Users as opposed to the "nt authority" account. WinGet does however expose COM APIs (both in-process and out-of-process). That's how the PowerShell cmdlets in the Microsoft.WinGet.Client module function in the system context.

You can logically install a packaged application in the system context, but it requires the registration step to function when a user logs into the system in order to "run" for the user. As I stated above, I believe there is more work to be done to make this more graceful when one is attempting to use WinGet in the system context to ensure packaged (MSIX) applications are provisioned such that every user will be registered when they log in.

For the benefit of other future time-travelers, WinGet was originally designed and built to help developers manage packages for their own needs (as a user). We needed time to build the product out and address many other concerns before it would be suitable as an enterprise IT solution. We're making progress on that front, and we will continue to do so. Many other products and services are leveraging WinGet today, and as it continues to mature, the product will get better.

**Reply by discotimetraveler** [reply]
> The notion of "machine wide" for an MSIX based package is handled differently in the sense that the virtual registry entries and the container the packaged application runs in doesn't get provisioned for the user until they log in.

Thankfully I got out of the Remote Desktop Services / Citrix business. But ironically this is half the reason I did so. Remote Desktop Services is a Microsoft technology yet Microsoft engineers are not developing software to be compatible with a core feature of it's [yes server] Operating system. To @cheesecracker's point, not only does Windows have to be configured for anyone who logs in, RDS has to be configured as if _everyone_ logging in is logging in for the first time. Because "Profile Management" in Citrix / Remote Desktop Services is a huge PITA. So _anything_ that installs to `%userprofile%` or more specifically `%appdata%` or `%localappdata%` is essentially incompatible with these modes of compute. We used to call it SBC back in the day, Server Based Computing. (When you have 500 Remote Desktop Servers and people log into a different one each day, it's the profile management software that maintains the user state so users feel like the system maintains their preferences and configurations).

This is essentially where MSIX installs, right? So, for conversations sake, if someone wanted Windows Terminal to run on a Citrix server, they'd have to use some overly complicated Profile Management software to "hopefully" get it to work. But now each individual user's profile contains a copy of software that should simply be sitting in C:\Program Files. Which not only increases storage requirements for profiles but also drastically complicates profile management. The more stuff that has to be shoved into a Profile Management tool, the worse they are and the worse experience a user of the system has.

Look, I get the reason why installing to Local Profile is "desired" for many products. For one it doesn't require UAC prompt or Admin rights and that's a WHOLE OTHER topic of suck for Windows. But to not provide the option for a "System Wide" install (meaning, not at all tied to user profile (except configuration state)) is just bad.

I didn't spend as much time on this topic as I could have, but I did make my post because I was trying to build a Windows Jump Box for a group of IT people to log into and use. And for one, getting Windows Terminal to even install on Server 2022 (?!?!?!) was hugely over-complicated (why??) and for two, when I asked my co-worker to log in and launch Windows Terminal he had no icon for it in his start menu. So, I'm not sure what "magic" you're referring to above with regards to installing MSIX system wide, but... it's not simple. If it's possible it would be nice to see some documentation on it.  Until then, this product will just remain an amazing tool for my client laptop but a trash tool for provisioning "en masse" to a group of people. ¯\\\_(ツ)\_/¯

**Reply by DrusTheAxe** [reply]
> This is essentially where MSIX installs, right?

Nope :-)

MSIX can install to a variety of types of locations e.g.

1. Under a PackageVolume e.g. `C:\Program Files\WindowsApps\...pkgfullname...`
2. System package under %WINDIR%\SystemApps e.g. `C:\Windows\SystemApps\Microsoft.Windows.ContentDeliveryManager_cw5n1h2txyewy`
3. Packaged installed with options.DevelopmentMode=true e.g. VisualStudio F5 registers a package in a project's `$(OutDir)`
4. options.StageInPlace=true e.g. used by AVD to mount pkgdir(s) in a VHD from a remote server
5. External Location

and a couple other cases I'm forgetting at the moment.

#1 is what you're probably thinking. Windows ships with a default PackageVolume (`C:\Program Files\WindowsApps`) but you can define more via APIs, Powershell cmdlets (`Add-AppxVolume`) and other means. In Settings `System > Storage > Where new content is saved` you can pick other drives for "New apps will save to:". If you pick not C: the PackageVolume API's called to define a new `PackageVolume` on the target volume (if it doesn't exist) and make it the default. `PackageManager` APIs optionally take a target `PackageVolume` and if not specified the default is used.

NOTE: A `PackageVolume` is a directory owned by Windows where Deployment can Stage packages (aka `mkdir and store package's files`). This location is heavily locked down so it can be shared across users -- doesn't matter how many users have a package registered, we only need 1 copy under a PackageVolume and _reliably_ can share across users. There's additional safeguards that can detect if a pkgdir's been modified -- any file added, removed or changed -- triggering remediation scenarios (access blocked, package content fixed/reacquired, etc).

This affords MSIX the opportunity to provide further optimizations, most notably de-duping disk space. Since pkgdirs are so locked down and reliably unmodified we not only share 1 pkgdir across all users registering the package, but if 1 file appears in 2 packages we can share that file's storage! When a file is laid down we check if the identical bytes are already present in another file and if so, and we can share, we hardlink the new file to the old one. Thus if 5 packages all ship the same file MSIX creates 5 file system entries but 1x disk blocks (not 5x). All 5 file system entries will have the same FILEID.

'and we can'. MSIX only does this hardlink optimization if the files' can physically share e.g. are on the same volume (can't hardlink across volumes), file's in a healthy status (PackageStatus=OK), and some other criteria.

This leads to significant disk savings. It also can speed up install times e.g. add a package foo-v2 to a system that already has foo-v1 and when laying down v2 any of its files which are identical to files in v1 will be de-duped. Hardlinked. No need to download the bytes from the package source (https: etc) or write the copy to disk.

This de-duping is inherent in MSIX' PackageVolume behavior. You don't need to do anything to enable it. ItJustWorks(TM). This is also why Windows recommends if you have files you don't need to change then don't e.g. if you compile SQLite to your own DLL and you update your package with the same version of SQLite don't recompile the SQLite dll. Even with identical source a dll/exe has timestamps in the PE file when it's built. Avoid doing this if you can to maximize the de-duping opportunities.

P.S. This is why Settings' Installed apps shows you the size you will free up if you uninstall a package, and not the size the package is currently using. For instance, if a package has 100 1M files and 20 are identical to files in other packages and are hardlinked/de-duped, `DIR ...pkgdir...` will show you the pkgdir's using 100M disk space but Settings will show you 80M will be freed up.

(And of course if a package is registered for 2 users and one 'Uninstalls' the package the pkgdir isn't removed. You'd free up whatever space was used by the user's data (e.g. `ApplicationData.LocalFolder`) but not the pkgdir itself)

VDI environments (AVD, Citrix, VMware, yadda yadda...) often use the [StageInPlace](https://learn.microsoft.com/uwp/api/windows.management.deployment.stagepackageoptions.stageinplace) property to mount a remote filesystem and tell Windows "stage the package at Q:\foo\bar with options.StageInPlace=true" via

> pkgmgr.StagePackageByUriAsync("file://q:/foo/bar", new StagePackageOptions(){ StageInPlace=true });

(or whatever the path). This tells Windows to use this directory as the package's installed location (NOT `C:\Program Files\WindowsApsps\...` or other PackageVolumes).

**Reply by DrusTheAxe** [reply]
> Windows Terminal installed from Microsoft Store is also user-wide only.

BTW that's not quite true. Terminal is installed by Store as an Add, but Terminal is typically provisioned on most all machines.

When a package family is provisioned and v5 is on the box every user gets v5 registered. When you Add v6 Windows is biased to highest version so it will now consider v6 the provisioned package and updates all users to it at the next opportunity (worst case their next login).

Net result? V6 is added to the system, users get updated to v6 at next opportunity and v5 gets automagically removed/deleted when no longer needed.

So Store updating Terminal "appears" to be doing a per-machine update but it's really not. It's doing a per-user update, but Windows' provisioning logic smartly assimilates updates when added to the machine.

See my comments above about provisioning and "all users"

**Comment by denelon** [comment]
> This is essentially where MSIX installs, right?

Not exactly. One of the benefits of the MSIX packaging format is the binary is only located in one place on the disk. When the user logs in and the package is registered, they do not get their own binary copy. It references the single physical copy installed on the system. I did a quick search on documentation covering this behavior and didn't find what I was looking for. I'll reach out internally to see if they can point me in the right direction to help here.

**Reply by DrusTheAxe** [reply]
This thread's meandering over and conflating multiple related but distinct concepts and behaviors. I've posted [above](https://github.com/microsoft/winget-cli/discussions/1983#discussioncomment-13043667) about the varying types of 'package locations' where an MSIX package's files can be staged, and how they're locked down enabling sharing across users and even potentially across packages.

On the 'per-user' vs 'per-machine' front it's somewhat complicated because terminology is often used loosely or informally. So let's be a little pedantic to clarify matters

The word 'install' isn't a proper noun in MSIX. We use it all the time but, technically, "installing" a package involves 3 key phases in the deployment pipeline:

* Index
* Stage
* Register

When you call `pkgmgr.AddPackageByUriAsync()` the package is run thru the deployment pipeline and hits all 3 of these stages.

* Indexing parses the incoming package info (appxmanifest.xml, signing, package path, yadda yadda) and stores the info.
* Staging put the package content on the box. Super-oversimplified, this is "MDIR pkgdir + Extract files from .msix to pkgdir"
* Register - Wire up the package to the user's profile

Indexing+Staging are done once per machine. Registration is per-user.

Indexing is implicit for any package entering the system, so we can ignore it for the rest of the conversation.

Deployment operations run through the 'deployment pipeline' which is some 50+ steps doing various things - staging package content in a pkgdir, verifying package signing, firing off notifications when the operation completes, and more. All the work is effectively a bucket-brigade of tasks handling all deployment operations (some tasks have conditional logic - depending on the request and machine state some tasks may be skip some work, or in some cases skip the entire task).

"Add" is effectively Stage + Register. Technically `pkgmgr.AddPackage*()` runs a request through the deployment pipeline which hits all stages including stage + register. You could call `pkgmgr.StagePackage*()` + `pkgmgr.RegisterPackage*()`  and accomplish the same net result -- `StagePackage*()` runs a request through the deployment pipeline but stops once the package staging handler is complete (the rest of the pipeline's not executed). Then call `RegisterPackage*()` and the request runs thru the deployment pipeline, but all the earlier tasks see their work's already done (Indexing, Staging, etc) so only the latter steps in the pipeline do any heavy lifting.

That's how a package is ultimately "installed" for a user -- Indexed and Staged on the machine, and Registered for that user. When a 2nd user "installs" a package it's already Indexed and Staged so the deployment pipeline only needs to do the Register work -- wiring up the user profile to the package (create start menu items, define file type associations, yadda yadda).

That's the "per-user" scenario.

If you want a package "installed" for all users you can "Provision" a package. Easiest way is to `pkgmgr.StagePackageByUriAsync()` and then `pkgmgr.ProvisionPackageForAllUsersAsync()`. The latter adds the package family to the "list of provisioned packages" tracked by Windows. When a user logs in, Windows compares what they have registered vs what they _should_ have and then updates them as necessary.

For example, if you have Calc v5 registered and at login v6 is on the machine Windows will update you to v6. Likewise if you have registered SomeObsoletePackageThatsBeenEndOfLifed then Windows will deregister that package at login.

`ProvisionPackageForAllUsersAsync` adds the package family to the list-of-provisioned-packages and also registers the package for the current user if necessary. Other users not logged in will have the package registered at their next login. Other users currently logged in and using the package will continue without impact, and only be switched to the newer version at the next opportunity (worst cast at their next login).

This differs from classic "per-machine" installation for updates where all running processes using the installed bits, across all users, must be stopped so the software can be updated. This is highly impactful, and why MSIX doesn't work this way. Instead, MSIX provides an "all users" model where a package is provisioned and users are moved to it if not currently in use, or will eventually move to it when not in use, with the worst cast at next login. The classic "per-machine" installation requires everyone in the machine be changed in lockstep, whereas MSIX follows more of a "all users coalesce to the newer version over time" for a less disruptive experience. Otherwise, classic "per-machine" and "all user" models are equivalent -- the software's available for use by all users current and future.

That explains how MSIX handles per-user and per-machine (all user) software deployment. I hope this helps clarify the misunderstandings and inaccuracies stated earlier in the thread.

**Reply by DrusTheAxe** [reply]
>anything that installs to %userprofile% or more specifically %appdata% or %localappdata% is essentially incompatible with these modes of compute.

Good thing MSIX doesn't :-)

MSIX does store **_data_** in the user's profile, and _**all**_ VDI environments have some challenges here. They often take a chainsaw approach to roaming profile data and often lack the nuance to not roam machine-specific data. For example, `ApplicationData.LocalCacheFolder` is where packaged applications can store data not meant for backup/restore, let alone roaming to another machine. VDI environments often roam the entire %USERPROFILE% including inappropriately grabbing these machine-affine locations.

But this is a classic VDI data problem. The bulk of the thread is about software (code, not data), for which MSIX is extremely VDI friendly.
